--- dwm.c	2016-06-04 15:52:46.460259017 -0400
+++ ./dwm.c	2016-06-04 15:50:31.116255801 -0400
@@ -768,16 +768,13 @@
 	drw_text(drw, x, 0, w, bh, m->ltsymbol, 0);
 	x += w;
 	xx = x;
-	if (m == selmon) { /* status is only drawn on selected monitor */
-		w = TEXTW(stext);
-		x = m->ww - w;
-		if (x < xx) {
-			x = xx;
-			w = m->ww - xx;
-		}
-		drw_text(drw, x, 0, w, bh, stext, 0);
-	} else
-		x = m->ww;
+	w = TEXTW(stext);
+	x = m->ww - w;
+	if(x < xx) {
+		x = xx;
+		w = m->ww - xx;
+	}
+	drw_text(drw, x, 0, w, bh, stext, 0);
 	if ((w = x - xx) > bh) {
 		x = xx;
 		if (m->sel) {
@@ -2042,9 +2039,11 @@
 void
 updatestatus(void)
 {
+	Monitor* m;
 	if (!gettextprop(root, XA_WM_NAME, stext, sizeof(stext)))
 		strcpy(stext, "dwm-"VERSION);
-	drawbar(selmon);
+	for(m = mons; m; m = m->next)
+		drawbar(m);
 }

 void
